File: matlab/feature_update.m
function particle= feature_update(particle, z, idf, R)
%function particle= feature_update(particle, z, idf, R)
%
% Having selected a new pose from the proposal distribution, this pose is assumed
% perfect and each feature update may be computed independently and without pose uncertainty.

xf= particle.xf(:,idf);
Pf= particle.Pf(:,:,idf);

[zp,Hv,Hf,Sf]= compute_jacobians(particle, idf, R);
v= z-zp; v(2,:)= pi_to_pi(v(2,:));

for i=1:length(idf)
    vi= v(:,i);
    Hfi= Hf(:,:,i);
    Pfi= Pf(:,:,i);
    xfi= xf(:,i);

    [xf(:,i), Pf(:,:,i)]= KF_cholesky_update(xfi,Pfi, vi,R,Hfi);
end

particle.xf(:,idf)= xf;
particle.Pf(:,:,idf)= Pf;

File: cpp/feature_update.cpp
#include "feature_update.h"
#include <iostream>

using namespace std;

void feature_update(Particle &particle, MatrixXf z, vector<int>idf, MatrixXf R)
{
    //Having selected a new pose from the proposal distribution, this pose is assumed perfect and each feature update maybe computed independently and without pose uncertainty

    int rows = (particle.xf()).rows();
    MatrixXf xf(rows,idf.size());    
    vector<MatrixXf> Pf;

    unsigned i,r; 
    for (i=0; i<idf.size(); i++) {
        for (r=0; r<(particle.xf()).rows(); r++) {
            xf(r,i) = (particle.xf())(r,(idf[i]));	
        }
        Pf.push_back((particle.Pf())[idf[i]]); //particle.Pf is a array of matrices
    }

    MatrixXf zp;    
    vector<MatrixXf> Hv;
    vector<MatrixXf> Hf;
    vector<MatrixXf> Sf;
    compute_jacobians(particle,idf,R,zp,&Hv,&Hf,&Sf);

	#if 0
    cout<<"after compute_jacobians"<<endl;    
    cout<<endl; 
    cout<<"in feature_update"<<endl;
    cout<<"zp is "<<endl;
    cout<<zp<<endl;
    cout<<"zp should be"<<endl;
    cout<<"25.6675 24.5294"<<endl;
    cout<<"-1.4925 0.1547"<<endl;
    cout<<endl;
	#endif

    MatrixXf v = z-zp; //TODO: need to fix: idf.size() is zero. which makes this break.

    unsigned c;
    for (c=0; c<v.cols();c++) {
        v(1,c) = pi_to_pi(v(1,c));
    }

    VectorXf vi(v.rows());
    MatrixXf Hfi;
    MatrixXf Pfi;
    VectorXf xfi(xf.rows());

    for (i=0; i< idf.size(); i++) {
        for (r=0; r<v.rows(); r++) {
            vi(r) = v(r,i); //v.column(i);
        }
        Hfi = Hf[i];  
        Pfi = Pf[i];
        for (r=0; r<xf.rows(); r++) {
            xfi(r) = xf(r,i); //xfi = xf.column(i);
        }
        KF_cholesky_update(xfi,Pfi,vi,R,Hfi);
        //TODO: insert xfi into xf, etc
        for (r=0; r<xf.rows(); r++) {
            xf(r,i) = xfi(r);
        }
        Pf[i] = Pfi; 
    }

    particle.setXf(xf);
    particle.setPf(Pf);
}
